ENH: sparse.linalg.spsolve: reduce memory footprint and increase speed

broesler opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

## What does this implement/fix?
This PR makes a substantial improvement to the peak memory usage and a reduction in run-time of `sparse.linalg.spsolve` when solving systems with many sparse right-hand sides. The original implementation looped over each column of the input, converted it to a dense column, and computed the full coordinates of the output with calls to `np.full`. The new implementation uses the `sparse.csc_array` format to avoid those full array creations.

This PR also introduces the `batch_size` kwarg that can be used to control the number of columns solved at one time. The `SuperLU` solver can handle dense matrix inputs natively at the C level, so instead of using a python loop over each column of the sparse RHS matrix input to `spsolve`, we can convert, *e.g.*, 100 columns at a time to dense and pass them to `SuperLU`. The input `batch_size=1` retains the original `spsolve` behavior. The `batch_size` argument does not apply to `use_umfpack=True` cases, because the low-level `scikit-umfpack` routines used in `spsolve` cannot handle matrix inputs.


## Additional information
[Benchmark results](https://broesler.github.io/scipy/#summarylist?sort=0&dir=asc) show a 35-50% reduction in peak memory use, with no change in run time for `batch_size=1`.

#### Peak memory use vs commit, `batch_size=1`
<img width="2531" height="1251" alt="image" src="https://github.com/user-attachments/assets/7ae693db-11c5-4ba3-a373-cf5f75bce086" />

#### Runtime vs commit, `batch_size=1`
<img width="2539" height="1250" alt="image" src="https://github.com/user-attachments/assets/9d918b53-1efe-4524-b690-eb7d3e480751" />

Note that the benchmarks across commits are only valid for `batch_size=1`, since that option was introduced in the latest commit.


### Default Batch Size
I chose the default batch size to optimize runtime and memory use. Plots of peak memory use and timing shows minimums for `batch_size=10` on my local machine. Peak memory use is not significantly changed vs. `batch_size=1`, but the benchmark shows a 50% reduction in runtime by using `batch_size=10`.

#### Peak memory use vs batch_size
<img width="2537" height="1244" alt="image" src="https://github.com/user-attachments/assets/6b7378ed-029b-4e98-992d-bf4bf24b8f80" />

#### Runtime vs commit vs batch_size
<img width="2532" height="1249" alt="image" src="https://github.com/user-attachments/assets/efea5594-6a02-4f2c-b9a3-db2b65c50e44" />

Memory and Runtime tables filtered to show only the `use_umfpack=False` results (SuperLU):

#### Peak Memory Usage (`peakmem_solve`)
*Lower is better. `batch_size=10` uses the least memory.*

| Nsq | K | Density | Batch=1 (Baseline) | Batch=10 | Batch=100 | Batch=5000 (Full) |
|---:|---:|---:|---:|---:|---:|---:|
| 5000 | 2000 | 0.01 | 331M | **318M** | 385M | 550M |
| 5000 | 2000 | 0.1 | 358M | **349M** | 413M | 606M |
| 5000 | 2000 | 0.5 | 501M | **487M** | 541M | 641M |

#### Execution Time (`time_solve`)
*Lower is better. `batch_size=10` is consistently the fastest.*

| Nsq | K | Density | Batch=1 (Baseline) | Batch=10 | Batch=100 | Batch=5000 (Full) |
|---:|---:|---:|---:|---:|---:|---:|
| 5000 | 2000 | 0.01 | 810±40ms | **391±3ms** | 441±8ms | 498±8ms |
| 5000 | 2000 | 0.1 | 843±40ms | **395±20ms** | 437±10ms | 491±4ms |
| 5000 | 2000 | 0.5 | 864±60ms | **423±30ms** | 458±5ms | 530±20ms |

broesler committed:
ENH: Rewrite sparse RHS loop in spsolve to directly build CSC array.

This commit eliminates the calls to `np.full` in favor of directly
computing the `indptr` array used by CSC format. It is more memory
efficient, especially for large, sparse RHSes.

broesler committed:
ENH: Rewrite sparse RHS loop in spsolve to directly build CSC array.

This commit eliminates the calls to `np.full` in favor of directly
computing the `indptr` array used by CSC format. It is more memory
efficient, especially for large, sparse RHSes.

broesler committed:
TST: Add warning filter for when has_umfpack=True.

This commit fixes a bug where a test fails due to
SparseEfficiencyWarning on non-CSC/CSR input, but otherwise passes. This
test is only run when scikit-umfpack is installed.

broesler committed:
TST: Add warning filter for when has_umfpack=True.

This commit fixes a bug where a test fails due to
SparseEfficiencyWarning on non-CSC/CSR input, but otherwise passes. This
test is only run when scikit-umfpack is installed.

broesler committed:
ENH: Use batched input for SuperLU spsolve with multiple RHS.

This commit takes advantage of the fact that the SuperLU module can take
an N x K RHS input and return an N x K output, without the need for
a python loop in spsolve. UMFPACK's low-level routines cannot handle
multiple RHS, so just set batch_size = 1 if the user requests
use_umfpack=True.

broesler committed:
ENH: Use batched input for SuperLU spsolve with multiple RHS.

This commit takes advantage of the fact that the SuperLU module can take
an N x K RHS input and return an N x K output, without the need for
a python loop in spsolve. UMFPACK's low-level routines cannot handle
multiple RHS, so just set batch_size = 1 if the user requests
use_umfpack=True.

broesler committed:
ENH: Write sparse RHS loop using blocks of csc_array and sparse.hstack.

broesler committed:
ENH: Write sparse RHS loop using blocks of csc_array and sparse.hstack.

broesler committed:
BENCH: Add benchmarks for sparse.linalg.spsolve.

broesler committed:
BENCH: Add benchmarks for sparse.linalg.spsolve.

broesler committed:
BENCH: Update spsolve benchmark to test batch_size kwarg.

broesler committed:
BENCH: Update spsolve benchmark to test batch_size kwarg.

lucascolley reviewed (commented):
Cool work! Just a quick comment on the parameter name: perhaps it should be specific to the RHS in some way? In case there are ever future developments which allowed you to pass multiple LHS matrices too. E.g. I'm working on that for the iterative solvers now. 

cc @dschmitz89 also

broesler committed:
MNT: Refactor batch_size -> rhs_batch_size.

This commit is in response to a reviewer comment to distinguish batch
processing of the RHS input from batch processing of ND arrays
(forthcoming).

broesler committed:
MNT: Refactor batch_size -> rhs_batch_size.

This commit is in response to a reviewer comment to distinguish batch
processing of the RHS input from batch processing of ND arrays
(forthcoming).

broesler committed:
MNT: Pre-allocate dense RHS buffer.

broesler committed:
MNT: Pre-allocate dense RHS buffer.

dschmitz89 commented:
LGTM but I cannot comment on the details as I have never done batched linear solves.

broesler committed:
BENCH: Update benchmark to emphasize large number RHS vs number rows.

This commit changes the parameters of the sparse.linalg.spsolve
benchmark to emphasize a large number of RHS relative to the number of
rows in the system.

broesler committed:
BENCH: Update benchmark to emphasize large number RHS vs number rows.

This commit changes the parameters of the sparse.linalg.spsolve
benchmark to emphasize a large number of RHS relative to the number of
rows in the system.

broesler committed:
ENH: Add input checking on rhs_batch_size.

broesler committed:
ENH: Add input checking on rhs_batch_size.

broesler committed:
TST: Add tests for rhs_batch_size.

broesler committed:
TST: Add tests for rhs_batch_size.

broesler commented:
I just rebased onto main and it fixed the remaining failing checks. Let me know if there is anything else that needs to be done before merging.

broesler committed:
DOC: Add versionadded tag to spsolve docstring for rhs_batch_size.

broesler committed:
DOC: Add versionadded tag to spsolve docstring for rhs_batch_size.

j-bowhay commented on scipy/sparse/linalg/_dsolve/tests/test_linsolve.py:
```diff
@@ -449,6 +451,29 @@ def test_bug_8278(self):
+        x = spsolve(A, b, rhs_batch_size=rhs_batch_size, use_umfpack=False)
+        if K == 1:
+            x_true = x_true.ravel()
+        assert_array_almost_equal(x, x_true)
```
Please can new tests use `assert_allclose` instead

broesler committed:
MNT: Make spsolve(..., rhs_batch_size) a keyword-only argument.

This commit is in response to a reviewer suggestion.

broesler committed:
MNT: Make spsolve(..., rhs_batch_size) a keyword-only argument.

This commit is in response to a reviewer suggestion.

broesler committed:
MNT: Use scipy utils to validate rhs_batch_size.

This commit is in response to a reviewer comment.

broesler committed:
MNT: Use scipy utils to validate rhs_batch_size.

This commit is in response to a reviewer comment.

broesler committed:
TST: Update test of bad_rhs_batch_size for _validate_int usage.

broesler committed:
TST: Update test of bad_rhs_batch_size for _validate_int usage.

broesler committed:
TST: Use assert_allclose for new tests.

This commit is in response to a reviewer comment.

broesler committed:
TST: Use assert_allclose for new tests.

This commit is in response to a reviewer comment.

broesler commented on scipy/sparse/linalg/_dsolve/linsolve.py:
```diff
@@ -251,6 +261,9 @@ def spsolve(A, b, permc_spec=None, use_umfpack=True):
 
     use_umfpack = use_umfpack and useUmfpack.u
 
+    if not isinstance(rhs_batch_size, Integral) or rhs_batch_size <= 0:
```
Thanks! I didn't realize `_validate_int` existed.

lucascolley reviewed (commented):
thanks @broesler, a few cursory comments.

By the way, would you happen to have any interest in batched iterative solvers? If so, you may be interested in reading https://discuss.scientific-python.org/t/batching-linearoperator-status-and-benchmarks/2199.

dschmitz89 commented:
> The pre-allocation of the `b_batch` array marginally improves peak memory and timing:
> 
> ### Peak Memory
> Nsq	K	Density	Batch	Before [[0b1b25d](https://github.com/scipy/scipy/commit/0b1b25dd58d0cc46a125ef48a0294466b587a26e)]	After [[452ef95](https://github.com/scipy/scipy/commit/452ef9547afa0d2617376f5addf021cd8c9117fe)]	Ratio
> 500	5003	0.01	1	127M	127M	1.00
> 500	5003	0.01	10	127M	127M	1.00
> 500	5003	0.01	100	131M	128M	0.98
> 500	5003	0.01	6000	212M	211M	0.99
> 500	5003	0.10	1	138M	137M	1.00
> 500	5003	0.10	10	132M	132M	1.00
> 500	5003	0.10	100	133M	133M	0.99
> 500	5003	0.10	6000	214M	213M	1.00
> 500	5003	0.50	1	160M	160M	1.00
> 500	5003	0.50	10	154M	154M	1.00
> 500	5003	0.50	100	159M	154M	0.97
> 500	5003	0.50	6000	240M	235M	0.98
> ### Timing
> Nsq	K	Density	Batch	Before [[0b1b25d](https://github.com/scipy/scipy/commit/0b1b25dd58d0cc46a125ef48a0294466b587a26e)]	After [[452ef95](https://github.com/scipy/scipy/commit/452ef9547afa0d2617376f5addf021cd8c9117fe)]	Ratio
> 500	5003	0.01	1	404 ± 1 ms	414 ± 0.8 ms	1.02
> 500	5003	0.01	10	97.1 ± 0.5 ms	89.4 ± 0.5 ms	0.92
> 500	5003	0.01	100	81.2 ± 3 ms	81.8 ± 4 ms	1.01
> 500	5003	0.01	6000	77.6 ± 1 ms	75.5 ± 2 ms	0.97
> 500	5003	0.10	1	406 ± 0.7 ms	416 ± 0.4 ms	1.02
> 500	5003	0.10	10	96.4 ± 0.5 ms	88.6 ± 0.5 ms	0.92
> 500	5003	0.10	100	83.5 ± 3 ms	82.5 ± 4 ms	0.99
> 500	5003	0.10	6000	77.6 ± 2 ms	75.7 ± 2 ms	0.98
> 500	5003	0.50	1	411 ± 0.6 ms	419 ± 0.7 ms	1.02
> 500	5003	0.50	10	98.5 ± 0.4 ms	89.9 ± 0.2 ms	0.91
> 500	5003	0.50	100	86.2 ± 4 ms	84.5 ± 4 ms	0.98
> 500	5003	0.50	6000	80.2 ± 1 ms	76.1 ± 1 ms	0.95
> I tweaked the benchmark to have a greater ratio of number of RHS to number of rows. The original results above still hold.
> 
> It turns out the bottleneck is actually the creation of a temporary `csc_array` for the slice, before converting to dense (whether or not in a pre-allocated buffer). It takes ~2-3x longer to create the slice than to convert the slice to dense (using `%timeit` tests). It would be neat if `csc_array` had a `.dense_slice()` method that avoided the intermediate array object, but that is for another PR.

The benchmarks do not show a clear performance improvement here. Do we need more benchmarks for the case of multiple right sides or batching?

broesler committed:
MNT/TST: Refactor rhs_batch_size -> rhs_block_size.

broesler committed:
MNT/TST: Refactor rhs_batch_size -> rhs_block_size.

tylerjereddy commented:
I'm not comfortable merging this right before branching -- if it was ready a domain expert should have merged it since the last comment on Feb. 19th. In the meanwhile, I'm going to bump the milestone -- I need to get a move on for branching.

broesler committed:
WIP: BENCH: Update benchmarks for ND batching.
